src/Functions/FunctionsHashing.h

```diff
@@ -1026,16 +1026,49 @@ class FunctionAnyHash : public IFunction
                 if constexpr (Impl::use_int_hash_for_pods)
                 {
                     if constexpr (std::is_same_v<ToType, UInt64>)
-                        h = IntHash64Impl::apply(bit_cast<UInt64>(vec_from[i]));
+                    {
+                        UInt64 v = bit_cast<UInt64>(vec_from[i]);
+                        if constexpr (std::endian::native == std::endian::big)
+                            v = __builtin_bswap64(v);
+                        h = IntHash64Impl::apply(v);
+                    }
                     else
-                        h = IntHash32Impl::apply(bit_cast<UInt32>(vec_from[i]));
+                    {
+                        UInt32 v = bit_cast<UInt32>(vec_from[i]);
+                        if constexpr (std::endian::native == std::endian::big)
+                            v = __builtin_bswap32(v);
+                        h = IntHash32Impl::apply(v);
+                    }
                 }
                 else
                 {
-                    if (std::is_same_v<Impl, JavaHashImpl>)
-                        h = JavaHashImpl::apply(vec_from[i]);
+                    if constexpr (std::endian::native == std::endian::little)
+                    {
+                        if (std::is_same_v<Impl, JavaHashImpl>)
+                            h = JavaHashImpl::apply(vec_from[i]);
+                        else
+                            h = apply(key, reinterpret_cast<const char *>(&vec_from[i]), sizeof(vec_from[i]));
+                    }
                     else
-                        h = apply(key, reinterpret_cast<const char *>(&vec_from[i]), sizeof(vec_from[i]));
+                    {
+                        if (std::is_same_v<Impl, JavaHashImpl>)
+                            h = JavaHashImpl::apply(vec_from[i]);
+                        else
+                        {
+                            if constexpr (std::is_same_v<ToType, UInt64>)
```

Comment:
Sure the distinction between 32 and 64 bit cases must be made based on `ToType`? We byte-reverse `vec_from[i]` (l. 1060) which is dependent on `FromType` (l. 1016, l. 1020).

Reply:
We distinguish 32 and 64 bit in order to use bit_cast<64> or bit_cast<32>, and then use __builtin_bswap64 or __builtin_bswap32 if it is big endian. Since from and to type can be of different size, it is pretty much natural way to make it right on both little and endian machines. Suppose we just use reverse memcpy, it is troublesome to calculate starting positions for copying.